v1/plugin/stream_collector_proxy.go

```diff
@@ -1,119 +1,179 @@
 package plugin
 
 import (
+	"errors"
 	"fmt"
+	"time"
 
 	"golang.org/x/net/context"
 
 	"github.com/intelsdi-x/snap-plugin-lib-go/v1/plugin/rpc"
 )
 
+const (
+	defaultMaxCollectDuration = 10 * time.Second
+	defaultMaxMetricsBuffer   = 0
+)
+
 type StreamProxy struct {
 	pluginProxy
-
 	plugin StreamCollector
+
+	// maxMetricsBuffer is the maximum number of metrics the plugin is buffering before sending metrics.
+	// Defaults to zero what means send metrics immediately.
+	maxMetricsBuffer int64
+
+	// maxCollectionDuration sets the maximum duration (always greater than 0s) between collections before metrics are sent.
+	// Defaults to 10s what means that after 10 seconds no new metrics are received, the plugin should send
+	// whatever data it has in the buffer instead of waiting longer.
+	maxCollectDuration time.Duration
+}
+
+func (p *StreamProxy) GetMetricTypes(ctx context.Context, arg *rpc.GetMetricTypesArg) (*rpc.MetricsReply, error) {
+	cfg := fromProtoConfig(arg.Config)
+
+	r, err := p.plugin.GetMetricTypes(cfg)
+	if err != nil {
+		return nil, err
+	}
+	metrics := []*rpc.Metric{}
+	for _, mt := range r {
+		// We can ignore this error since we are not returning data from GetMetricTypes.
+		metric, _ := toProtoMetric(mt)
+		metrics = append(metrics, metric)
+	}
+	reply := &rpc.MetricsReply{
+		Metrics: metrics,
+	}
+	return reply, nil
 }
 
-func (c *StreamProxy) StreamMetrics(stream rpc.StreamCollector_StreamMetricsServer) error {
+func (p *StreamProxy) StreamMetrics(stream rpc.StreamCollector_StreamMetricsServer) error {
+	if stream == nil {
+		return errors.New("Stream metrics server is nil")
+	}
+
 	// Error channel where we will forward plugin errors to snap where it
 	// can report/handle them.
 	errChan := make(chan string)
 	// Metrics into the plugin from snap.
 	inChan := make(chan []Metric)
 	// Metrics out of the plugin into snap.
 	outChan := make(chan []Metric)
-	err := c.plugin.StreamMetrics(inChan, outChan, errChan)
+
+	err := p.plugin.StreamMetrics(inChan, outChan, errChan)
 	if err != nil {
 		return err
 	}
-	go metricSend(c.plugin, outChan, stream)
-	go errorSend(c.plugin, errChan, stream)
-	streamRecv(c.plugin, inChan, stream)
+
+	go p.metricSend(outChan, stream)
+	go p.errorSend(errChan, stream)
+	p.streamRecv(inChan, stream)
```

Comment:
**To reviewers**:

All those functions are now StreamProxy methods what is necessary to pass maxCollectDuration and maxMetricsBuffer. As the result, `c.plugin` does not have to be passed as input to them either.